pkg/hubble/container/ring_test.go

```diff
@@ -91,6 +93,79 @@ func BenchmarkTimeLibRingRead(b *testing.B) {
 	})
 }
 
+func BenchmarkReadBufferNoReaders(b *testing.B) {
+	benchmarkRingBuffer(b, benchmarkRingBufferOptions{})
+}
+
+func BenchmarkReadBufferOneReader(b *testing.B) {
+	benchmarkRingBuffer(b, benchmarkRingBufferOptions{
+		readers: 1,
+	})
+}
+
+func BenchmarkReadBufferEightReaders(b *testing.B) {
+	b.Skip() // FIXME
+	benchmarkRingBuffer(b, benchmarkRingBufferOptions{
+		readers: 8,
+	})
+}
+
+type benchmarkRingBufferOptions struct {
+	capacity capacity
+	readers  int
+}
+
+// benchmarkRingBuffer benchmarks RingBuffer's writes and reads with multiple
+// readers.
+func benchmarkRingBuffer(b *testing.B, options benchmarkRingBufferOptions) {
+	b.ReportAllocs()
+
+	if options.capacity == capacity(0) {
+		options.capacity = benchmarkCapacity
+	}
+
+	r := NewRing(options.capacity)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	readerEventsReceived := make(chan int, options.readers)
+	for i := 0; i < options.readers; i++ {
+		rr := NewRingReader(r, 0)
+		go func() {
+			// Count the number of events received and report them over a
+			// channel.
+			eventsReceived := 0
+			for {
+				if rr.NextFollow(ctx) == nil {
+					readerEventsReceived <- eventsReceived
+					return
+				}
+				eventsReceived++
+			}
+		}()
+	}
+
+	event := &v1.Event{}
+
+	b.ResetTimer()
+
+	// Write the same event b.N times.
+	for i := 0; i < b.N; i++ {
+		r.Write(event)
+	}
+
+	// Cancel all readers.
+	cancel()
```

Comment:
This is racy and probably the reason why no event gets actually read. `NextFollow(ctx)` waits on an event or for the context to be done (when the context is done, `<nil>` is returned).

If you actually add something like `<-time.After(1*time.Second)` just before the `cancel()` call, the test actually succeeds (provided that `b.N+1` events have been written, see comment above).

Reply:
This is benchmarking code so any `<-time.After(1*time.Second)` (or the equivalent `time.Sleep(time.Second)`)) will mess up the benchmarks sadly.

Reply:
I was not suggesting to add `<-time.After(1*time.Second)` to the test. I mentioned it to indicate that when adding this, the tests passes meaning that there's some synchronization logic problem in your code.

Reply:
Sorry I misunderstood. The problem is that with the current implementation there is no way to synchronize readers. Without the `time.Sleep` all the events are written before the goroutines in the readers are scheduled, so by the time the readers execute there are no events left for them to receive. With a long-enough `time.Sleep` (at least one scecond on my machine) the reader goroutines do get scheduled and we observe some events.